dags/inference/maxtext_inference_microbenchmark.py

```diff
@@ -0,0 +1,244 @@
+  for key_value_axis_order_product_id in range(
+      len(key_value_cache_idx_product_idx_values)
+  ):
+    (
+        key_axis_order_idx,
+        value_axis_order_idx,
+    ) = key_value_cache_idx_product_idx_values[
+        int(key_value_axis_order_product_id)
+    ]
+    ar_key_axis_order_str = cache_permu_idx_strs[key_axis_order_idx]
+    ar_value_axis_order_str = cache_permu_idx_strs[value_axis_order_idx]
+
+    key_value_axis_order_product_id_list.append(key_value_axis_order_product_id)
+    ar_key_axis_order_str_list.append(ar_key_axis_order_str)
+    ar_value_axis_order_str_list.append(ar_value_axis_order_str)
+
+  key_value_axis_order_product_id_split = numpy.array_split(
+      key_value_axis_order_product_id_list, sweep_vm_count
+  )
+  ar_key_axis_order_str_split = numpy.array_split(
+      ar_key_axis_order_str_list, sweep_vm_count
+  )
+  ar_value_axis_order_str_split = numpy.array_split(
+      ar_value_axis_order_str_list, sweep_vm_count
+  )
+  key_value_axis_order_product_id_concat_list = [
+      ":".join(list(str(y) for y in x))
+      for x in key_value_axis_order_product_id_split
+  ]
+  prefill_key_axis_order_concat_list = [
+      ":".join(list(x)) for x in ar_key_axis_order_str_split
+  ]
+  prefill_value_axis_order_concat_list = [
+      ":".join(list(x)) for x in ar_key_axis_order_str_split
+  ]
+  ar_key_axis_order_concat_list = [
+      ":".join(list(x)) for x in ar_value_axis_order_str_split
+  ]
+  ar_value_axis_order_concat_list = [
+      ":".join(list(x)) for x in ar_value_axis_order_str_split
+  ]
+
+  return (
+      key_value_axis_order_product_id_concat_list,
+      prefill_key_axis_order_concat_list,
+      prefill_value_axis_order_concat_list,
+      ar_key_axis_order_concat_list,
+      ar_value_axis_order_concat_list,
+  )
+
+
+# Run once a day at 12 pm UTC (4 am PST)
+SCHEDULED_TIME = "0 12 * * *" if composer_env.is_prod_env() else None
+
+
+with models.DAG(
+    dag_id="maxtext_inference_microbenchmark",
+    schedule=SCHEDULED_TIME,
+    tags=[
+        "inference_team",
+        "maxtext",
+        "nightly",
+        "benchmark",
+        "microbenchmark",
+    ],
+    start_date=datetime.datetime(2024, 1, 19),
+    catchup=False,
+) as dag:
+  sweep_vm_count = 12
+  (
+      key_value_axis_order_product_id_concat_list,
+      prefill_key_axis_order_concat_list,
+      prefill_value_axis_order_concat_list,
+      ar_key_axis_order_concat_list,
+      ar_value_axis_order_concat_list,
+  ) = get_concatenated_list_of_params(sweep_vm_count=sweep_vm_count)
+  test_name_prefix = "maxtext-inf-microbench"
+  test_models = {
+      "llama2-7b": {
+          "tpu_version_cores": [(TpuVersion.V5E, 4)],
+          "base_output_directory": "gs://inference-benchmarks/logs/llama2-7b/microbenchmark/int8/tmp/",
+          "tokenizer": "tokenizer.llama2",
+          "weight_dtype": "bfloat16",
+          "inference_microbenchmark_prefill_lengths": 1024,
+          "inference_microbenchmark_stages": "generate",
```

Comment:
Can you do both prefill and generate? something like -- ["prefill", "generate", "prefill, generate"] ?

Reply:
For any nightly or routine runs, we can simply do "prefill,generate". For specific tuning runs, we may want to run "prefill" only or "generate" only. Different run scenarios will yield different metrics.

Reply:
Sounds good. Let me do a test run with both "prefill,generate"

Reply:
When I run both prefill and generate, and flatten the outputs, it looks like this. Since the original dict is nested, I have to do recursively generate prefixes based on the parent key. What do you think? Do
```
{'Prefill_Result_prefill_length_num_prefill_logits_params': 32000, 'Prefill_Result_prefill_length_total_prefill_logits_size': 128000, 'Prefill_Result_prefill_length_avg_prefill_logits_param_size': 4.0, 'Prefill_Result_prefill_length_num_prefill_cache_params': 541130784, 'Prefill_Result_prefill_length_total_prefill_cache_size': 545521792, 'Prefill_Result_prefill_length_avg_prefill_cache_param_size': 1.0081145041639323, 'Prefill_1024_prefill_time_in_ms': 23.2722, 'Prefill_1024_prefill_total_tflops_per_device': 3.519484526592, 'Prefill_1024_prefill_tflops_per_sec_per_device': 151.23127708562146, 'Prefill_Insert_1024_prefill_insert_time_in_ms': 23.7819, 'AutoRegressive_ar_step_in_ms': 76.6489, 'AutoRegressive_ar_step_in_ms_per_seq': 0.7984260416666666, 'AutoRegressive_ar_global_batch_size': 96, 'AutoRegressive_ar_total_throughput_tokens_per_second': 1252.4641579983536, 'AutoRegressive_ar_device_bandwidth_GB_per_second': 193.65452560963044, 'sizes_Model_size_in_GB': 7.0035456, 'sizes_cache_size_in_GB': 52.370079872, 'sizes_model_params_in_billions': 6.739775488}
```